test/extensions/filters/http/ext_authz/ext_authz_integration_test.cc

```diff
@@ -0,0 +1,218 @@
+
+    auto* attributes = check_request.mutable_attributes();
+    auto* http_request = attributes->mutable_request()->mutable_http();
+
+    // Clear fields which are not relevant.
+    attributes->clear_source();
+    attributes->clear_destination();
+    attributes->clear_metadata_context();
+    http_request->clear_id();
+    http_request->clear_headers();
+
+    EXPECT_EQ(check_request.DebugString(), expected_check_request.DebugString());
+
+    result = ext_authz_request_->waitForEndStream(*dispatcher_);
+    RELEASE_ASSERT(result, result.message());
+  }
+
+  void waitForSuccessfulUpstreamResponse() {
+    AssertionResult result =
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_);
+    RELEASE_ASSERT(result, result.message());
+    result = upstream_request_->waitForEndStream(*dispatcher_);
+    RELEASE_ASSERT(result, result.message());
+
+    upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+    upstream_request_->encodeData(response_size_, true);
+    response_->waitForEndStream();
+
+    EXPECT_TRUE(upstream_request_->complete());
+    EXPECT_EQ(request_body_.length(), upstream_request_->bodyLength());
+
+    EXPECT_TRUE(response_->complete());
+    EXPECT_EQ("200", response_->headers().Status()->value().getStringView());
+    EXPECT_EQ(response_size_, response_->body().size());
+  }
+
+  void sendExtAuthzResponse() {
+    ext_authz_request_->startGrpcStream();
+    envoy::service::auth::v2::CheckResponse check_response;
+    check_response.mutable_status()->set_code(Grpc::Status::GrpcStatus::Ok);
+    ext_authz_request_->sendGrpcMessage(check_response);
+    ext_authz_request_->finishGrpcStream(Grpc::Status::Ok);
+  }
+
+  const std::string expectedRequestBody() {
+    const std::string request_body_string = request_body_.toString();
+    const uint64_t request_body_length = request_body_.length();
+    return request_body_length > max_request_bytes_
+               ? request_body_string.substr(0, max_request_bytes_)
+               : request_body_string;
+  }
+
+  void cleanup() {
+    if (fake_ext_authz_connection_ != nullptr) {
+      if (clientType() != Grpc::ClientType::GoogleGrpc) {
+        AssertionResult result = fake_ext_authz_connection_->close();
+        RELEASE_ASSERT(result, result.message());
+      }
+      AssertionResult result = fake_ext_authz_connection_->waitForDisconnect();
+      RELEASE_ASSERT(result, result.message());
+    }
+    cleanupUpstreamAndDownstream();
+  }
+
+  void expectCheckRequestWithBody(Http::CodecClient::Type downstream_protocol,
+                                  uint64_t request_size) {
+    initializeWithDownstreamProtocol(downstream_protocol);
+    initiateClientConnection(request_size);
+    waitForExtAuthzRequest(fmt::format(R"EOF(
```

Comment:
@dio The integration tests are great but is there a way that we can make this part a bit more readable?

Reply:
Sure. Updated.